feat!: admission controller Helm chart consolidation (#1802)
kubewarden/kubewarden-controller: charts/kubewarden-controller/README.md

# Kubewarden Admission Controller

Unified Helm chart for deploying the complete Kubewarden admission control stack.

> > **Note:** This chart combines what were previously three separate charts:

Reviewer: I would drop this note or mention that it only applies since `1.37`.
But I would prefer to rename the chart to a new chart `adm-controller` altogether.

Author: Yeah, let's rename the whole helm-chart in a follow up PR. Just to avoid more changes in this big PR.

Reviewer 2: fine with me

> `kubewarden-crds` (CRDs), `kubewarden-controller` (controller), and
> `kubewarden-defaults` (default PolicyServer and recommended policies).

## Installation

```sh
helm install kubewarden kubewarden/kubewarden-controller -n kubewarden --create-namespace
```

> ## Migration from Three-Chart Setup

Reviewer: I would separate this guidance into its own `docs/migrations.md` file or similar. We can add a section there for 1.36 to 1.37 charts.

Author: I'm not sure about this. This README.md file is the one embedded in the Helm chart tarball. I would say that it's nice to have there to allow user to have a minimum of documentation about this migration path there. What do you think?

Reviewer 2: I think we can leave that in the README and remove that in 4 releases from now

> If you're currently running the legacy three-chart setup (`kubewarden-crds`, `kubewarden-controller`, `kubewarden-defaults`), follow these steps to migrate to the unified chart.

Reviewer: This needs to specify the versions we are migrating from and to.

**⚠️ Important**: There will be a brief window during migration where no admission control is active. Plan accordingly.

### Prerequisites

- Access to your cluster with `kubectl` and `helm`
- Backup tool or `kubectl` configured

### Migration Steps

#### 1. Backup All Policies and PolicyServers

> Uninstalling `kubewarden-crds` cascade-deletes **all** custom resources, so every policy and PolicyServer must be backed up:

Reviewer: Given that since 1.36 we can now delete the policy-server `default`, and uninstall controller and still have all the policies in the cluster, I would:

1. Ensure you have 1.36 installed, as it keeps the policies around even if the controller and default PolicyServer are removed.
2. The user collects the 3 values.yaml from the 3 charts, and concatenates them together into a `values-1.37.yml`.
3. Annotate the CRD templates with a `helm.sh/resource-policy: keep`, so they are kept on uninstall. Instead of doing a manual `kubectl label`, I would release a `kubewarden-crds-1.36.1` that sets this by default (or with a `migration.keep-crds: true` values key that can do that for us and instruct the user to install and enable it). This means that we need to release such a chart; we may need to not merge this PR or open this PR later on against a feat branch before continuing.
4. Once the CRD templates in the cluster are marked as `keep`, uninstall `kubewarden-crds`. The CRDs stick around, and so will all PolicyServers and policies definitions.
5. Install the new `kubewarden-controller-1.37` (or `adm-controller-1.37`) with `helm install --take-ownership --values values-137.yml`  which will take ownership of the CRDs.
6. Done.

Author: Ok, I think this approach is doable. But it requires an additional change. The annotation only is not enough. We need to remove the [pre-delete-hook](https://github.com/kubewarden/adm-controller/blob/0af49d28d746a8837498393b787a215516896682/charts/kubewarden-controller/templates/pre-delete-hook.yaml#L39) from the `kubewarden-controller` as well. This hook deletes all the policy servers in the cluster. Therefore, it will delete the custom policy server from the users. Thus, the path would be:
- Release `kubewarden-crds` adding the `helm.sh/resource-policy="keep"` annotation
- Release `kubewarden-controller` removing/updating the `pre-delete-hook.yaml` hook to delete only the resources owned by the chart

Once these helm charts are released and the users update to them, they will be able to migrate to the single helm chart by following these steps:

- Uninstall `kubewarden-defaults`
- Uninstall `kubewarden-controller`
- Uninstall `kubewarden-crds`

At this stage, users have their own policies and policy servers running but all the default resources should be gone. They then can continue

- Merge the previous values files or pass them all to the next helm install command
- Install the merged `adm-controller` helm chart using the previous values and with the `--take-ownership` CLI flag. The CLI flag  will transfer  CRDs  ownership to the new `adm-controller` installation by updating the helm annotations

After that, the process finishes when the new reconciler starts and recreates the default resources from the new `ConfigMap` where they are defined. As long as the values used at install time match the previous ones, the resulting resources will be equivalent to those from the old defaults chart.

Furthermore, during my tests I try to avoid the patch release to add the annotation in the CRDs and update in the hooks. However, Helm ignores the annotation added by `kubectl` it considers only the its stored definition. Therefore, the release is required. We could avoid the `kubewarden-controller` asking the user to use the `--no-hooks` CLI flag. But as we already need to release the CRDs chart, let's update the controller as well. I think it's less error prone for users.

<details>
<summary>Script used to test the procedure</summary>

```bash
#!/usr/bin/env bash
# test-migration.sh
#
# Exercises the PR #1802 migration path proposed in
# https://github.com/kubewarden/adm-controller/pull/1802#discussion_r3380239545
#
# Flow:
#   1. (Re)create a kind cluster.
#   2. Install the legacy three-chart stack (kubewarden-crds,
#      kubewarden-controller, kubewarden-defaults) pinned to v1.36.0.
#   3. Deploy a CUSTOM PolicyServer + a ClusterAdmissionPolicy
#      (pod-privileged) bound to it.
#   4. Probe admission: a non-privileged pod is admitted, a privileged pod
#      is rejected.
#   5. Annotate the CRDs with helm.sh/resource-policy=keep, simulating a
#      hypothetical kubewarden-crds-1.36.1 release.
#   6. helm uninstall the three legacy releases in reverse order.
#   7. Best-effort mid-migration probe.
#   8. helm install the unified 1.37 chart from the working tree with
#      --take-ownership and locally-built v1.37.0 images.
#   9. Re-probe admission on the surviving CUSTOM policy.
#
# The kind cluster is intentionally left running on exit. Re-running the
# script removes the leftover cluster at phase 2.

set -euo pipefail

#------------------------------------------------------------------------------
# Configuration (overridable via environment)
#------------------------------------------------------------------------------
CLUSTER_NAME="${CLUSTER_NAME:-kubewarden-migration-test}"
KW_NAMESPACE="${KW_NAMESPACE:-kubewarden}"

LEGACY_APP_VERSION="${LEGACY_APP_VERSION:-v1.36.0}"
LEGACY_CRDS_CHART_VERSION="${LEGACY_CRDS_CHART_VERSION:-1.28.0}"
LEGACY_CONTROLLER_CHART_VERSION="${LEGACY_CONTROLLER_CHART_VERSION:-5.14.0}"
LEGACY_DEFAULTS_CHART_VERSION="${LEGACY_DEFAULTS_CHART_VERSION:-3.14.0}"

NEW_CHART_PATH="${NEW_CHART_PATH:-./charts/kubewarden-controller}"
NEW_RELEASE_NAME="${NEW_RELEASE_NAME:-kubewarden}"
NEW_TAG="${NEW_TAG:-v1.37.0}"

CONTROLLER_IMAGE="${CONTROLLER_IMAGE:-ghcr.io/kubewarden/adm-controller/controller:${NEW_TAG}}"
POLICY_SERVER_IMAGE="${POLICY_SERVER_IMAGE:-ghcr.io/kubewarden/adm-controller/policy-server:${NEW_TAG}}"
AUDIT_SCANNER_IMAGE="${AUDIT_SCANNER_IMAGE:-ghcr.io/kubewarden/adm-controller/audit-scanner:${NEW_TAG}}"

CUSTOM_POLICY_SERVER="${CUSTOM_POLICY_SERVER:-custom}"
CUSTOM_POLICY_NAME="${CUSTOM_POLICY_NAME:-custom-no-privileged-pod}"
POD_PRIVILEGED_MODULE="${POD_PRIVILEGED_MODULE:-registry://ghcr.io/kubewarden/policies/pod-privileged:v1.0.10}"

PROBE_NS="${PROBE_NS:-migration-probe}"

WAIT_TIMEOUT="${WAIT_TIMEOUT:-5m}"

LOG_FILE="${LOG_FILE:-./test-migration.log}"

KW_CRDS=(
  policyservers.policies.kubewarden.io
  clusteradmissionpolicies.policies.kubewarden.io
  admissionpolicies.policies.kubewarden.io
  clusteradmissionpolicygroups.policies.kubewarden.io
  admissionpolicygroups.policies.kubewarden.io
)

#------------------------------------------------------------------------------
# Helpers
#------------------------------------------------------------------------------
: > "$LOG_FILE"
exec > >(tee -a "$LOG_FILE") 2>&1

step() { printf '\n\033[1;36m==> %s\033[0m\n' "$*"; }
info() { printf '    %s\n' "$*"; }
fail() { printf '\n\033[1;31mFAIL: %s\033[0m\n' "$*" >&2; exit 1; }
ok()   { printf '    \033[1;32mOK\033[0m %s\n' "$*"; }

require_cmd() {
  command -v "$1" >/dev/null 2>&1 || fail "missing required command: $1"
}

helm_version_at_least_318() {
  local v
  v="$(helm version --short 2>/dev/null | sed -E 's/^v?([0-9]+)\.([0-9]+)\..*/\1 \2/' | head -n1)"
  local major minor
  read -r major minor <<<"$v"
  [[ -n "${major:-}" && -n "${minor:-}" ]] || return 1
  if (( major > 3 )); then return 0; fi
  if (( major == 3 && minor >= 18 )); then return 0; fi
  return 1
}

# Probe: create a Pod manifest in PROBE_NS, expect admission to either succeed
# or fail. Args: <name> <privileged true|false> <expect admit|reject>
probe_pod() {
  local name="$1" privileged="$2" expect="$3"
  local manifest
  manifest="$(cat <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: $name
  namespace: $PROBE_NS
spec:
  restartPolicy: Never
  containers:
  - name: c
    image: registry.k8s.io/pause:3.10
    securityContext:
      privileged: $privileged
EOF
)"
  # Idempotent: ensure no leftover from a previous run with same name.
  kubectl -n "$PROBE_NS" delete pod "$name" --ignore-not-found --wait=false >/dev/null 2>&1 || true

  local out rc
  out="$(printf '%s' "$manifest" | kubectl apply -f - 2>&1)" && rc=0 || rc=$?

  case "$expect" in
    admit)
      if (( rc == 0 )); then
        ok "pod '$name' admitted as expected"
      else
        info "kubectl output: $out"
        fail "pod '$name' was rejected but should have been admitted"
      fi
      ;;
    reject)
      if (( rc != 0 )); then
        ok "pod '$name' rejected as expected"
        info "rejection: $(printf '%s' "$out" | tail -n1)"
      else
        fail "pod '$name' was admitted but should have been rejected"
      fi
      ;;
    either)
      if (( rc == 0 )); then
        ok "pod '$name' admitted (best-effort probe)"
      else
        ok "pod '$name' rejected (best-effort probe): $(printf '%s' "$out" | tail -n1)"
      fi
      ;;
    *)
      fail "probe_pod: unknown expectation '$expect'"
      ;;
  esac
}

wait_policy_active() {
  local kind="$1" name="$2"
  info "waiting for $kind/$name to reach status.policyStatus=active"
  local i status
  for i in $(seq 1 60); do
    status="$(kubectl get "$kind" "$name" -o jsonpath='{.status.policyStatus}' 2>/dev/null || true)"
    if [[ "$status" == "active" ]]; then
      ok "$kind/$name is active"
      return 0
    fi
    sleep 5
  done
  kubectl get "$kind" "$name" -o yaml || true
  fail "$kind/$name did not reach active status within timeout"
}

#------------------------------------------------------------------------------
# Phase 1: preflight
#------------------------------------------------------------------------------
step "Phase 1: preflight"
for c in kind kubectl helm docker yq; do require_cmd "$c"; done

helm_version_at_least_318 || fail "helm >= 3.18 is required (--take-ownership). Found: $(helm version --short)"
ok "helm $(helm version --short) supports --take-ownership"

helm repo list 2>/dev/null | grep -q '^kubewarden' \
  || fail "the 'kubewarden' helm repo is not configured. Add with: helm repo add kubewarden https://charts.kubewarden.io"
helm repo update kubewarden >/dev/null
ok "kubewarden helm repo refreshed"

for img in "$CONTROLLER_IMAGE" "$POLICY_SERVER_IMAGE" "$AUDIT_SCANNER_IMAGE"; do
  docker image inspect "$img" >/dev/null 2>&1 \
    || fail "image not found in local docker: $img"
  ok "found local image $img"
done

[[ -d "$NEW_CHART_PATH" ]] || fail "new chart directory not found: $NEW_CHART_PATH"
ok "new chart directory: $NEW_CHART_PATH"

#------------------------------------------------------------------------------
# Phase 2: (re)create kind cluster
#------------------------------------------------------------------------------
step "Phase 2: kind cluster '$CLUSTER_NAME'"
if kind get clusters 2>/dev/null | grep -qx "$CLUSTER_NAME"; then
  info "cluster '$CLUSTER_NAME' already exists; deleting it first"
  kind delete cluster --name "$CLUSTER_NAME"
fi
kind create cluster --name "$CLUSTER_NAME"
kubectl cluster-info --context "kind-${CLUSTER_NAME}" >/dev/null
ok "cluster '$CLUSTER_NAME' is up"

#------------------------------------------------------------------------------
# Phase 3: preload v1.37.0 images
#------------------------------------------------------------------------------
step "Phase 3: preload v1.37.0 images into kind"
for img in "$CONTROLLER_IMAGE" "$POLICY_SERVER_IMAGE" "$AUDIT_SCANNER_IMAGE"; do
  kind load docker-image --name "$CLUSTER_NAME" "$img"
  ok "loaded $img"
done

#------------------------------------------------------------------------------
# Phase 4: install legacy 1.36 stack
#------------------------------------------------------------------------------
step "Phase 4: install legacy ${LEGACY_APP_VERSION} three-chart stack"
kubectl create namespace "$KW_NAMESPACE" >/dev/null 2>&1 || true

helm upgrade --install kubewarden-crds kubewarden/kubewarden-crds \
  --namespace "$KW_NAMESPACE" \
  --version "$LEGACY_CRDS_CHART_VERSION" \
  --wait --timeout "$WAIT_TIMEOUT"
ok "kubewarden-crds installed"

helm upgrade --install kubewarden-controller kubewarden/kubewarden-controller \
  --namespace "$KW_NAMESPACE" \
  --version "$LEGACY_CONTROLLER_CHART_VERSION" \
  --wait --timeout "$WAIT_TIMEOUT"
ok "kubewarden-controller installed"

helm upgrade --install kubewarden-defaults kubewarden/kubewarden-defaults \
  --namespace "$KW_NAMESPACE" \
  --version "$LEGACY_DEFAULTS_CHART_VERSION" \
  --set "recommendedPolicies.enabled=true" \
  --set "recommendedPolicies.defaultPolicyMode=protect" \
  --wait --timeout "$WAIT_TIMEOUT"
ok "kubewarden-defaults installed (recommended policies enabled, mode=protect)"

step "Phase 4b: snapshot recommended ClusterAdmissionPolicies"
RECOMMENDED_POLICIES=()
while IFS= read -r name; do
  [[ -n "$name" ]] && RECOMMENDED_POLICIES+=("$name")
done < <(
  kubectl get clusteradmissionpolicies \
    -o jsonpath='{range .items[?(@.metadata.annotations.meta\.helm\.sh/release-name=="kubewarden-defaults")]}{.metadata.name}{"\n"}{end}'
)
(( ${#RECOMMENDED_POLICIES[@]} > 0 )) \
  || fail "no recommended ClusterAdmissionPolicies found from kubewarden-defaults release"
ok "found ${#RECOMMENDED_POLICIES[@]} recommended policies: ${RECOMMENDED_POLICIES[*]}"

for p in "${RECOMMENDED_POLICIES[@]}"; do
  wait_policy_active clusteradmissionpolicy "$p"
done

#------------------------------------------------------------------------------
# Phase 5: provision custom PolicyServer + policy
#------------------------------------------------------------------------------
step "Phase 5: provision custom PolicyServer + ClusterAdmissionPolicy"

cat <<EOF | kubectl apply -f -
apiVersion: policies.kubewarden.io/v1
kind: PolicyServer
metadata:
  name: $CUSTOM_POLICY_SERVER
spec:
  image: ghcr.io/kubewarden/adm-controller/policy-server:${LEGACY_APP_VERSION}
  replicas: 1
EOF

cat <<EOF | kubectl apply -f -
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: $CUSTOM_POLICY_NAME
spec:
  policyServer: $CUSTOM_POLICY_SERVER
  module: $POD_PRIVILEGED_MODULE
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations: ["CREATE"]
  mutating: false
  mode: protect
EOF

wait_policy_active clusteradmissionpolicy "$CUSTOM_POLICY_NAME"

#------------------------------------------------------------------------------
# Phase 6: pre-migration admission probe
#------------------------------------------------------------------------------
step "Phase 6: pre-migration admission probe"
kubectl create namespace "$PROBE_NS" >/dev/null 2>&1 || true
probe_pod pre-good  false admit
probe_pod pre-bad   true  reject

#------------------------------------------------------------------------------
# Phase 7: simulate kubewarden-crds-1.36.1 (chart with keep annotation)
#
# IMPORTANT: Helm 4 reads helm.sh/resource-policy=keep from the *stored*
# release manifest, NOT from the live cluster object. A plain
# `kubectl annotate` is therefore insufficient. We mimic the hypothetical
# `kubewarden-crds-1.36.1` release by pulling the legacy chart, injecting
# the annotation into every CRD template, and running `helm upgrade`. The
# upgrade rewrites the release manifest so the next `helm uninstall` honors
# the annotation.
#------------------------------------------------------------------------------
step "Phase 7: simulate kubewarden-crds-1.36.1 via helm upgrade with patched chart"

PATCHED_CHART_DIR="$(mktemp -d)/kubewarden-crds-patched"
trap '[[ -n "${PATCHED_CHART_DIR:-}" ]] && rm -rf "$(dirname "$PATCHED_CHART_DIR")"' EXIT
helm pull kubewarden/kubewarden-crds --version "$LEGACY_CRDS_CHART_VERSION" \
  --untar --untardir "$(dirname "$PATCHED_CHART_DIR")"
mv "$(dirname "$PATCHED_CHART_DIR")/kubewarden-crds" "$PATCHED_CHART_DIR"

info "injecting helm.sh/resource-policy: keep into CRD templates"
for f in "$PATCHED_CHART_DIR"/templates/crds/policies.kubewarden.io_*.yaml; do
  yq -i '.metadata.annotations."helm.sh/resource-policy" = "keep"' "$f"
  v="$(yq '.metadata.annotations."helm.sh/resource-policy"' "$f")"
  [[ "$v" == "keep" ]] || fail "annotation injection failed for $f"
  ok "patched $(basename "$f")"
done

helm upgrade kubewarden-crds "$PATCHED_CHART_DIR" \
  --namespace "$KW_NAMESPACE" \
  --wait --timeout "$WAIT_TIMEOUT"
ok "kubewarden-crds upgraded with keep annotation in stored manifest"

info "verifying annotation is present on the live CRDs"
for crd in "${KW_CRDS[@]}"; do
  v="$(kubectl get crd "$crd" -o jsonpath='{.metadata.annotations.helm\.sh/resource-policy}')"
  [[ "$v" == "keep" ]] || fail "annotation not applied on live CRD $crd"
  ok "$crd has helm.sh/resource-policy=keep"
done

#------------------------------------------------------------------------------
# Phase 8: uninstall legacy releases (reverse order)
#------------------------------------------------------------------------------
step "Phase 8: uninstall legacy releases in reverse order"
# IMPORTANT: the legacy kubewarden-controller chart ships a pre-delete Job
# that runs `kubectl delete --all policyservers.policies.kubewarden.io`,
# which would wipe every PolicyServer in the cluster (including user-owned
# ones). We pass --no-hooks to bypass it. This is another caveat for the
# migration documentation.
helm uninstall kubewarden-defaults   --namespace "$KW_NAMESPACE" --wait || true
helm uninstall kubewarden-controller --namespace "$KW_NAMESPACE" --wait --no-hooks || true
helm uninstall kubewarden-crds       --namespace "$KW_NAMESPACE" --wait || true

info "verifying CRDs survived"
for crd in "${KW_CRDS[@]}"; do
  kubectl get crd "$crd" >/dev/null 2>&1 || fail "CRD $crd was deleted; keep annotation did not work"
  ok "$crd still present"
done

info "verifying user CRs survived"
kubectl get policyserver "$CUSTOM_POLICY_SERVER" >/dev/null 2>&1 \
  || fail "custom PolicyServer was deleted"
kubectl get clusteradmissionpolicy "$CUSTOM_POLICY_NAME" >/dev/null 2>&1 \
  || fail "custom ClusterAdmissionPolicy was deleted"
ok "custom PolicyServer and ClusterAdmissionPolicy survived"

#------------------------------------------------------------------------------
# Phase 8a: mid-migration probe (best effort)
#------------------------------------------------------------------------------
step "Phase 8a: mid-migration admission probe (best effort)"
info "controller is gone; webhook configurations and custom policy-server pod should still serve"
probe_pod mid-good false either
probe_pod mid-bad  true  either

#------------------------------------------------------------------------------
# Phase 9: install unified 1.37 chart with --take-ownership
#------------------------------------------------------------------------------
step "Phase 9: install unified chart with --take-ownership"

helm install "$NEW_RELEASE_NAME" "$NEW_CHART_PATH" \
  --namespace "$KW_NAMESPACE" \
  --take-ownership \
  --set "image.repository=kubewarden/adm-controller/controller" \
  --set "image.tag=${NEW_TAG}" \
  --set "auditScanner.image.repository=kubewarden/adm-controller/audit-scanner" \
  --set "auditScanner.image.tag=${NEW_TAG}" \
  --set "policyServer.image.repository=kubewarden/adm-controller/policy-server" \
  --set "policyServer.image.tag=${NEW_TAG}" \
  --set "recommendedPolicies.enabled=false" \
  --wait --timeout "$WAIT_TIMEOUT"
ok "unified chart installed"

info "verifying CRDs are now owned by the new release"
for crd in "${KW_CRDS[@]}"; do
  rel="$(kubectl get crd "$crd" -o jsonpath='{.metadata.annotations.meta\.helm\.sh/release-name}')"
  [[ "$rel" == "$NEW_RELEASE_NAME" ]] \
    || fail "CRD $crd is not owned by release '$NEW_RELEASE_NAME' (got: '$rel')"
  ok "$crd owned by $NEW_RELEASE_NAME"
done

#------------------------------------------------------------------------------
# Phase 10: post-migration verification
#------------------------------------------------------------------------------
step "Phase 10: post-migration verification"

info "waiting for the custom ClusterAdmissionPolicy to reconcile under the new controller"
wait_policy_active clusteradmissionpolicy "$CUSTOM_POLICY_NAME"

info "verifying recommended-policy CRs from legacy kubewarden-defaults are gone"
for p in "${RECOMMENDED_POLICIES[@]}"; do
  if kubectl get clusteradmissionpolicy "$p" >/dev/null 2>&1; then
    fail "recommended policy '$p' still exists; it should have been cascade-deleted with kubewarden-defaults"
  fi
  ok "recommended policy '$p' is gone"
done

info "post-migration admission probe"
probe_pod post-good false admit
probe_pod post-bad  true  reject

step "Migration test PASSED"
info "kind cluster '$CLUSTER_NAME' is left running for inspection"
info "log saved to $LOG_FILE"

```

</details>

Author: I've update the README file with the migration documentation considering that we release the patch version of the current Helm charts.

Reviewer: Thanks for doing all this investigation! I think this is a good approach, as we leave the migration as gitos gitops as possible; just install and uninstall charts.

Given that we are going with doing a last patch release pre-migration, I wonder if we can do a one-up, and perform the migration without having policy downtime.

Is it possible to do the same process for the controller, with `helm.sh/resource-policy: keep `, `helm uninstall kubewarden-controller`, `helm install adm-controller`? Maybe with that we could pull it of.

Also, this would need to work for downstream chart consumers too.

Reviewer 2: What if, instead of releasing a new patch release of 1.36, we update the migration instructions to have the user manually add the "helm keep" annotation by hand?

Author: Manual annotation is not possible. I've already test that:

> Furthermore, during my tests I try to avoid the patch release to add the annotation in the CRDs and update in the hooks. However, Helm ignores the annotation added by kubectl it considers only the its stored definition. Therefore, the release is required. We could avoid the kubewarden-controller asking the user to use the --no-hooks CLI flag. But as we already need to release the CRDs chart, let's update the controller as well. I think it's less error prone for users.

For example, we can see in the current `kubewarden-crds` helm chart

```console
jvanz@localhost:~/SUSE/kubewarden> helm version
version.BuildInfo{Version:"v4.2.0", GitCommit:"06468084e85c244c712834933d25ea232a4c2093", GitTreeState:"clean", GoVersion:"go1.26.3", KubeClientVersion:"v1.36"}
jvanz@localhost:~/SUSE/kubewarden> helm install -n kubewarden --create-namespace --wait kubewarden-crds kubewarden/kubewarden-crds
NAME: kubewarden-crds
LAST DEPLOYED: Mon Jun 15 15:59:49 2026
NAMESPACE: kubewarden
STATUS: deployed
REVISION: 1
DESCRIPTION: Install complete
TEST SUITE: None
NOTES:
Kubewarden CRDs now available:
  clusteradmissionpolicies.policies.kubewarden.io
  admissionpolicies.policies.kubewarden.io
  clusteradmissionpolicygroups.policies.kubewarden.io
  admissionpolicygroups.policies.kubewarden.io
  policyservers.policies.kubewarden.io

OpenReport CRDs now available:
  reports.openreports.io
  clusterreports.openreports.io
jvanz@localhost:~/SUSE/kubewarden> kubectl annotate crd \
>    admissionpolicies.policies.kubewarden.io \
>    admissionpolicygroups.policies.kubewarden.io \
>    clusteradmissionpolicies.policies.kubewarden.io \
>    clusteradmissionpolicygroups.policies.kubewarden.io \
>    policyservers.policies.kubewarden.io \
>    helm.sh/resource-policy=keep --overwrite
customresourcedefinition.apiextensions.k8s.io/admissionpolicies.policies.kubewarden.io annotated
customresourcedefinition.apiextensions.k8s.io/admissionpolicygroups.policies.kubewarden.io annotated
customresourcedefinition.apiextensions.k8s.io/clusteradmissionpolicies.policies.kubewarden.io annotated
customresourcedefinition.apiextensions.k8s.io/clusteradmissionpolicygroups.policies.kubewarden.io annotated
customresourcedefinition.apiextensions.k8s.io/policyservers.policies.kubewarden.io annotated

jvanz@localhost:~/SUSE/kubewarden> k8s get crds policyservers.policies.kubewarden.io -o json | jq ".metadata.annotations"
{
  "controller-gen.kubebuilder.io/version": "v0.18.0",
  "helm.sh/resource-policy": "keep",
  "meta.helm.sh/release-name": "kubewarden-crds",
  "meta.helm.sh/release-namespace": "kubewarden"
}
jvanz@localhost:~/SUSE/kubewarden> helm uninstall -n kubewarden kubewarden-crds
release "kubewarden-crds" uninstalled
jvanz@localhost:~/SUSE/kubewarden> k8s get crds policyservers.policies.kubewarden.io -o json | jq ".metadata.annotations"
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "policyservers.policies.kubewarden.io" not found
```

```sh
kubectl get clusteradmissionpolicies -A -o yaml > clusteradmissionpolicies-backup.yaml
kubectl get admissionpolicies -A -o yaml > admissionpolicies-backup.yaml
kubectl get clusteradmissionpolicygroups -A -o yaml > clusteradmissionpolicygroups-backup.yaml
kubectl get admissionpolicygroups -A -o yaml > admissionpolicygroups-backup.yaml
kubectl get policyservers -o yaml > policyservers-backup.yaml

> ```

Reviewer 2: We should not use raw output of `kubectl get` since this will include fields that can cause troubles at restore time (metadata.uid, metadata.resourceVersion) or just introduce noise (metadata.managedFields, status).

Doing a quick research, and some prompting with LLM:

--- BEGIN (I cannot make a suggestion)

To prevent kubectl apply errors when restoring backups, you must strip cluster-specific metadata (like uid, resourceVersion, managedFields, and status) from the exported YAML.

Choose one of the following methods to safely back up your Kubewarden Custom Resources.

##### Using `kubectl-neat`

```console

# Install neat via krew if you haven't already:
# kubectl krew install neat

kubectl get clusteradmissionpolicies -A -o yaml | kubectl neat > clusteradmissionpolicies-backup.yaml
kubectl get admissionpolicies -A -o yaml | kubectl neat > admissionpolicies-backup.yaml
kubectl get clusteradmissionpolicygroups -A -o yaml | kubectl neat > clusteradmissionpolicygroups-backup.yaml
kubectl get admissionpolicygroups -A -o yaml | kubectl neat > admissionpolicygroups-backup.yaml
kubectl get policyservers -A -o yaml | kubectl neat > policyservers-backup.yaml
```

##### Using `yq`

```console
FILTER='del(.items[].metadata.uid, .items[].metadata.resourceVersion, .items[].metadata.creationTimestamp, .items[].metadata.generation, .items[].metadata.managedFields, .items[].status)'

kubectl get clusteradmissionpolicies -A -o yaml | yq "$FILTER" > clusteradmissionpolicies-backup.yaml
kubectl get admissionpolicies -A -o yaml | yq "$FILTER" > admissionpolicies-backup.yaml
kubectl get clusteradmissionpolicygroups -A -o yaml | yq "$FILTER" > clusteradmissionpolicygroups-backup.yaml
kubectl get admissionpolicygroups -A -o yaml | yq "$FILTER" > admissionpolicygroups-backup.yaml
kubectl get policyservers -A -o yaml | yq "$FILTER" > policyservers-backup.yaml
```

#### 2. Uninstall Old Charts

Uninstall in reverse order:

```sh
helm uninstall kubewarden-defaults -n kubewarden
helm uninstall kubewarden-controller -n kubewarden
helm uninstall kubewarden-crds -n kubewarden
```

This removes all CRDs and cascades deletion of all CRs (PolicyServers and policies).

#### 3. Install the Unified Chart

```sh
helm install kubewarden kubewarden/kubewarden-controller -n kubewarden
```

This creates:
- CRDs in `templates/crds/` (with `helm.sh/resource-policy: keep` to prevent deletion on uninstall)
- The controller

> - The default PolicyServer and recommended policies (if enabled)

Reviewer 2: ```suggestion
This defines:

- CRDs. These are created with the `helm.sh/resource-policy: keep` to prevent deletion on uninstall.
- The actual adm-controller
- If enabled by `values.yaml`, the default Policy Server and the recommended policies
```

#### 4. Restore User Policies

Once the default PolicyServer is ready, re-apply all backed-up resources:

```sh
kubectl apply -f policyservers-backup.yaml
kubectl apply -f clusteradmissionpolicies-backup.yaml
kubectl apply -f admissionpolicies-backup.yaml
kubectl apply -f clusteradmissionpolicygroups-backup.yaml
kubectl apply -f admissionpolicygroups-backup.yaml
```

> The controller's DefaultsApplier will overwrite any managed defaults with the correct ownership labels on the next reconciliation.

Reviewer 2: Maybe this is too low level to be mentioned inside of a README

Reviewer: I agree, the end user doesn't need to be bothered with these implementation details (that can change).

## Configuration

### Defaults

> The chart can deploy a default PolicyServer and recommended policies managed by the controller:

Reviewer 2: ```suggestion
The chart can deploy a default Policy Server and a series of recommended policies:
```

```yaml
defaultConfigMapName: kubewarden-defaults

policyServer:
  enabled: true
  replicaCount: 1
  # ... (see values.yaml for full options)

recommendedPolicies:
  enabled: false  # Disabled by default
  defaultPolicyMode: "monitor"
  allowPrivilegeEscalationPolicy:
    # ... (see values.yaml)
```

> When `policyServer.enabled: false` and `recommendedPolicies.enabled: false`, the defaults ConfigMap is not rendered and the controller cleans up all managed resources.

Reviewer 2: ```suggestion
**Note:**  these resources are owned and reconciled by the adm-controller.
Manual changes are going to be reverted. Also, changing this value to `false` leads to a cleanup of all these managed resources.
```

### CRDs

> CRDs are installed in `templates/crds/` with the `helm.sh/resource-policy: keep` annotation. This means:

Reviewer 2: ```suggestion
CRDs are installed with the `helm.sh/resource-policy: keep` annotation. This means:
```

- `helm upgrade` will update CRDs
- `helm uninstall` will **not** delete CRDs (preventing catastrophic cascade-deletion of all cluster resources)

To fully remove CRDs after uninstall:

```sh
kubectl delete crd policyservers.policies.kubewarden.io
kubectl delete crd clusteradmissionpolicies.policies.kubewarden.io
kubectl delete crd admissionpolicies.policies.kubewarden.io
kubectl delete crd clusteradmissionpolicygroups.policies.kubewarden.io
kubectl delete crd admissionpolicygroups.policies.kubewarden.io

> ```

Reviewer 2: Maybe this section should be moved to the next one (the "Uninstall" one).

## Uninstall

```sh
helm uninstall kubewarden -n kubewarden
```

This removes:
- The controller deployment
- Managed defaults (resources with `kubewarden.io/managed-by=kubewarden-controller-defaults` label)
- ConfigMaps, Secrets, Services

It does **not** remove:
- CRDs (due to `helm.sh/resource-policy: keep`)
- User-managed PolicyServers and policies

## Version

- Chart version: 6.0.0-alpha.1

> - App version: v2.0.0-alpha.1

Reviewer 2: I would remove this section, it's just a PITA to keep it up to date.

## References

- [Kubewarden Documentation](https://docs.kubewarden.io/)

> - [RFC 0026: Unified Admission Controller Chart](https://github.com/kubewarden/rfc/blob/main/rfc/0026-unified-admission-controller-chart.md)

Reviewer 2: I would remove that, users don't care about these technical docs. I would instead bring back the link to the releases page.

name must match the old one or Kubernetes rejects the update due to
immutable selectors. Getting any step wrong can delete resources or
break admission, so the script handles it all and verifies each step.

What survives the migration:

- The five Kubewarden CRDs.
- Your custom `PolicyServer` instances and policy CRs, along with
  their `Validating`/`MutatingWebhookConfiguration` objects.
- The `default` `PolicyServer` and recommended policy CRs (if
  enabled). The unified chart's DefaultsApplier adopts them in
  place. The `policy-server-default` Deployment is owned by the
  `PolicyServer` CR through an owner reference, so as long as the
  CR survives, the Deployment stays up and admission continues.
- The `kubewarden-ca` Secret. Already-running policy-server pods
  stay trusted by the new webhook CA bundles. No TLS rotation
  happens.

### Prerequisites

- Helm v4+ (needed for Server-Side Apply and post-renderer plugins)
- kubectl with access to your cluster
- yq v4 (github.com/mikefarah/yq) for the post-renderer
- jq for detecting installed chart versions
- The three legacy releases must be installed in your cluster

> ### What the script does

Reviewer 2: Somewhere I would still put a big fat warning recommending to take a backup of the custom policies and custom policy servers being deployed

Author: Done at https://github.com/kubewarden/adm-controller/pull/1802/commits/ae5c5e2bb743813cdf291668fd4ba217347ac445
